In [ ]:
import os
os.chdir("C:\\Users\\lopez\\Desktop\\udea-ai-4-eng-20251-pruebas-saber-pro-colombia") # Ruta donde se encuentra la carpeta con los archivos
os.listdir()

['submission_example.csv', 'test.csv', 'train.csv']

In [36]:
import pandas as pd
df = pd.read_csv("train.csv")
df.head()

ID  PERIODO         ESTU_PRGM_ACADEMICO ESTU_PRGM_DEPARTAMENTO  \
0  904256    20212                  ENFERMERIA                 BOGOTÁ   
1  645256    20212                     DERECHO              ATLANTICO   
2  308367    20203       MERCADEO Y PUBLICIDAD                 BOGOTÁ   
3  470353    20195  ADMINISTRACION DE EMPRESAS              SANTANDER   
4  989032    20212                  PSICOLOGIA              ANTIOQUIA   

             ESTU_VALORMATRICULAUNIVERSIDAD ESTU_HORASSEMANATRABAJA  \
0  Entre 5.5 millones y menos de 7 millones       Menos de 10 horas   
1  Entre 2.5 millones y menos de 4 millones                       0   
2  Entre 2.5 millones y menos de 4 millones         Más de 30 horas   
3  Entre 4 millones y menos de 5.5 millones                       0   
4  Entre 2.5 millones y menos de 4 millones     Entre 21 y 30 horas   

  FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET                 FAMI_EDUCACIONPADRE  \
0            Estrato 3                 Si    Técnica o tecnológica incompleta   
1            Estrato 3                 No      Técnica o tecnológica completa   
2            Estrato 3                 Si  Secundaria (Bachillerato) completa   
3            Estrato 4                 Si                             No sabe   
4            Estrato 3                 Si                   Primaria completa   

  FAMI_TIENELAVADORA  ... ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO  \
0                 Si  ...                     N                       No   
1                 Si  ...                     N                       No   
2                 Si  ...                     N                       No   
3                 Si  ...                     N                       No   
4                 Si  ...                     N                       No   

  FAMI_TIENECOMPUTADOR FAMI_TIENEINTERNET.1  \
0                   Si                   Si   
1                   Si                   No   
2                   No                   Si   
3                   Si                   Si   
4                   Si                   Si   

                  FAMI_EDUCACIONMADRE RENDIMIENTO_GLOBAL coef_1  coef_2  \
0                           Postgrado         medio-alto  0.322   0.208   
1    Técnica o tecnológica incompleta               bajo  0.311   0.215   
2  Secundaria (Bachillerato) completa               bajo  0.297   0.214   
3  Secundaria (Bachillerato) completa               alto  0.485   0.172   
4                   Primaria completa         medio-bajo  0.316   0.232   

   coef_3  coef_4  
0   0.310   0.267  
1   0.292   0.264  
2   0.305   0.264  
3   0.252   0.190  
4   0.285   0.294  

[5 rows x 21 columns]

# Rellenar valores null en la columna

Primero verificamos los valores null

In [37]:
print(df.isnull().sum()) # Verificacion numerica de valores null en todo el documento

df[df["FAMI_ESTRATOVIVIENDA"].isnull()] # verificacion en tabla de valores null FAMI_ESTRATOVIVIENDA

ID                                    0
PERIODO                               0
ESTU_PRGM_ACADEMICO                   0
ESTU_PRGM_DEPARTAMENTO                0
ESTU_VALORMATRICULAUNIVERSIDAD     6287
ESTU_HORASSEMANATRABAJA           30857
FAMI_ESTRATOVIVIENDA              32137
FAMI_TIENEINTERNET                26629
FAMI_EDUCACIONPADRE               23178
FAMI_TIENELAVADORA                39773
FAMI_TIENEAUTOMOVIL               43623
ESTU_PRIVADO_LIBERTAD                 0
ESTU_PAGOMATRICULAPROPIO           6498
FAMI_TIENECOMPUTADOR              38103
FAMI_TIENEINTERNET.1              26629
FAMI_EDUCACIONMADRE               23664
RENDIMIENTO_GLOBAL                    0
coef_1                                0
coef_2                                0
coef_3                                0
coef_4                                0
dtype: int64


ID  PERIODO                 ESTU_PRGM_ACADEMICO  \
21      284373    20195              INGENIERIA MECATRONICA   
26      925456    20195                             DERECHO   
49      836859    20195   LICENCIATURA EN CIENCIAS SOCIALES   
55      663184    20195                    INGENIERIA CIVIL   
58      629999    20212                      TRABAJO SOCIAL   
...        ...      ...                                 ...   
692444  763147    20183                          OPTOMETRIA   
692448   54207    20183                          PSICOLOGIA   
692454  821212    20195  ADMINISTRACION PUBLICA TERRITORIAL   
692464  592258    20183                 FILOLOGIA E IDIOMAS   
692487  350508    20195                  CONTADURÍA PÚBLICA   

       ESTU_PRGM_DEPARTAMENTO            ESTU_VALORMATRICULAUNIVERSIDAD  \
21                      VALLE  Entre 5.5 millones y menos de 7 millones   
26                    BOLIVAR  Entre 2.5 millones y menos de 4 millones   
49                     BOGOTÁ    Entre 1 millón y menos de 2.5 millones   
55                     TOLIMA  Entre 2.5 millones y menos de 4 millones   
58                      VALLE    Entre 1 millón y menos de 2.5 millones   
...                       ...                                       ...   
692444                 BOGOTÁ  Entre 4 millones y menos de 5.5 millones   
692448                 TOLIMA    Entre 1 millón y menos de 2.5 millones   
692454                 BOGOTÁ                                       NaN   
692464                 BOGOTÁ         Entre 500 mil y menos de 1 millón   
692487                 BOGOTÁ                                       NaN   

       ESTU_HORASSEMANATRABAJA FAMI_ESTRATOVIVIENDA FAMI_TIENEINTERNET  \
21                         NaN                  NaN                NaN   
26                         NaN                  NaN                NaN   
49                         NaN                  NaN                NaN   
55         Entre 21 y 30 horas                  NaN                 Si   
58                         NaN                  NaN                NaN   
...                        ...                  ...                ...   
692444                     NaN                  NaN                NaN   
692448                     NaN                  NaN                NaN   
692454                     NaN                  NaN                NaN   
692464                     NaN                  NaN                NaN   
692487                     NaN                  NaN                NaN   

                     FAMI_EDUCACIONPADRE FAMI_TIENELAVADORA  ...  \
21                                   NaN                NaN  ...   
26                                   NaN                NaN  ...   
49                                   NaN                NaN  ...   
55      Técnica o tecnológica incompleta                 Si  ...   
58                                   NaN                NaN  ...   
...                                  ...                ...  ...   
692444                 Primaria completa                NaN  ...   
692448                               NaN                NaN  ...   
692454                               NaN                NaN  ...   
692464                               NaN                NaN  ...   
692487                               NaN                NaN  ...   

       ESTU_PRIVADO_LIBERTAD ESTU_PAGOMATRICULAPROPIO FAMI_TIENECOMPUTADOR  \
21                         N                       No                  NaN   
26                         N                       No                  NaN   
49                         N                       Si                  NaN   
55                         N                       No                   Si   
58                         N                       No                  NaN   
...                      ...                      ...                  ...   
692444                     N                       No                  NaN   
692448                  

Se observa que tenemos 32137 valores null para la columna FAMI_ESTRATOVIVIENDA

Rellenamos los valores null de acuerdo a la moda, ya que es un valor categorico

In [38]:
df['FAMI_ESTRATOVIVIENDA'] = df['FAMI_ESTRATOVIVIENDA'].fillna(df['FAMI_ESTRATOVIVIENDA'].mode()[0]) # Rellenamos los valores null con la moda

df['FAMI_ESTRATOVIVIENDA'].head(50)

0     Estrato 3
1     Estrato 3
2     Estrato 3
3     Estrato 4
4     Estrato 3
5     Estrato 5
6     Estrato 2
7     Estrato 2
8     Estrato 1
9     Estrato 5
10    Estrato 1
11    Estrato 2
12    Estrato 3
13    Estrato 1
14    Estrato 3
15    Estrato 3
16    Estrato 1
17    Estrato 2
18    Estrato 1
19    Estrato 2
20    Estrato 2
21    Estrato 2
22    Estrato 1
23    Estrato 2
24    Estrato 2
25    Estrato 3
26    Estrato 2
27    Estrato 3
28    Estrato 3
29    Estrato 2
30    Estrato 2
31    Estrato 3
32    Estrato 2
33    Estrato 3
34    Estrato 4
35    Estrato 3
36    Estrato 4
37    Estrato 2
38    Estrato 2
39    Estrato 2
40    Estrato 4
41    Estrato 3
42    Estrato 2
43    Estrato 1
44    Estrato 2
45    Estrato 3
46    Estrato 2
47    Estrato 3
48    Estrato 3
49    Estrato 2
Name: FAMI_ESTRATOVIVIENDA, dtype: object

Finalmente podemos observar que en los valores 21 y 49 que anteriormente eran valores null, ahora tienen valores de acuerdo a la moda

In [39]:
print(df.isnull().sum())

ID                                    0
PERIODO                               0
ESTU_PRGM_ACADEMICO                   0
ESTU_PRGM_DEPARTAMENTO                0
ESTU_VALORMATRICULAUNIVERSIDAD     6287
ESTU_HORASSEMANATRABAJA           30857
FAMI_ESTRATOVIVIENDA                  0
FAMI_TIENEINTERNET                26629
FAMI_EDUCACIONPADRE               23178
FAMI_TIENELAVADORA                39773
FAMI_TIENEAUTOMOVIL               43623
ESTU_PRIVADO_LIBERTAD                 0
ESTU_PAGOMATRICULAPROPIO           6498
FAMI_TIENECOMPUTADOR              38103
FAMI_TIENEINTERNET.1              26629
FAMI_EDUCACIONMADRE               23664
RENDIMIENTO_GLOBAL                    0
coef_1                                0
coef_2                                0
coef_3                                0
coef_4                                0
dtype: int64


# Codificación etiqueta FAMI_ESTRATOVIVIENDA (One-Hot Encoding)

In [35]:
df["FAMI_ESTRATOVIVIENDA"].unique() #Valores en columna

array(['Estrato 3', 'Estrato 4', 'Estrato 5', 'Estrato 2', 'Estrato 1',
       'Estrato 6', 'Sin Estrato'], dtype=object)

In [ ]:
df["FAMI_ESTRATOVIVIENDA_encoded"] = pd.Categorical(
    df["FAMI_ESTRATOVIVIENDA"],
    categories=["Sin Estrato", "Estrato 1", "Estrato 2", "Estrato 3", "Estrato 4", "Estrato 5", "Estrato 6"],
    ordered=True,
).codes

df[["FAMI_ESTRATOVIVIENDA", "FAMI_ESTRATOVIVIENDA_encoded"]].drop_duplicates().sort_values("FAMI_ESTRATOVIVIENDA_encoded")


FAMI_ESTRATOVIVIENDA  FAMI_ESTRATOVIVIENDA_encoded
21                   NaN                            -1
176          Sin Estrato                             0
8              Estrato 1                             1
6              Estrato 2                             2
0              Estrato 3                             3
3              Estrato 4                             4
5              Estrato 5                             5
154            Estrato 6                             6